### DISCUSSION TITLE
initdb ICU provider

### DISCUSSION THREAD

**Discussion by eshikhov** [cloudnative-pg/cloudnative-pg#5163] [Q&A] [answered]
Hi!

is it possible to initialize the database with the parameters initdb --locale-provider=icu --icu-locale=und-x-icu?

**Comment by elafarge** [comment]
Well, @francoisjehl and I had a quick look at [the code](https://github.com/cloudnative-pg/cloudnative-pg/blob/main/api/v1/cluster_types.go#L1662-L1667) because we're facing the same problem.

It seems there's a deprecated `options` field in the `initdb` spec you can use:

```
  bootstrap:
    initdb: 
      options:
        - --encoding=UTF8
        - --locale-provider=icu
        - --icu-locale=und-x-icu
```

Given it's deprecated, it may not be such a good idea to use it (it seems to do exactly what you want though). 

I'll give a try at making a PR to add the `localeProvider` and `icuLocale` fields in the spec directly, no ETA on when I think I might finish and even less on when this will be merged (or if it will be merged at all).

**Reply by markmaker** [reply]
> It seems there's a deprecated `options` field in the `initdb` spec you can use:
> 
> ```
>   bootstrap:
>     initdb: 
>       options:
>         - --encoding=UTF8
>         - --locale-provider=icu
>         - --icu-locale=und-x-icu
> ```
> 
> Given it's deprecated, it may not be such a good idea to use it (it seems to do exactly what you want though).

I really, really need this, so I tried your workaround with `options`. It doesn't work for some reason (extracted messages): 

> Using language tag "x-icu" for ICU locale "und-x-icu"
> **initdb: error: locale "x-icu" has unknown language "x-icu"**

```
Defaulted container "initdb" out of: initdb, bootstrap-controller (init)
{"level":"info","ts":"2024-11-27T20:04:41.950308215Z","msg":"Creating new data directory","logging_pod":"test-1-initdb","pgdata":"/var/lib/postgresql/data/pgdata","initDbOptions":["--username","postgres","-D","/var/lib/postgresql/data/pgdata","--data-checksums","--encoding=UTF8","--lc-collate=C","--lc-ctype=C","--locale-provider=icu","--icu-locale=und-x-icu"]}
{"level":"info","ts":"2024-11-27T20:04:41.968028035Z","logger":"initdb","msg":"The files belonging to this database system will be owned by user \"postgres\".\nThis user must also own the server process.\n\nUsing language tag \"x-icu\" for ICU locale \"und-x-icu\".\n","pipe":"stdout","logging_pod":"test-1-initdb"}
{"level":"info","ts":"2024-11-27T20:04:41.968080497Z","logger":"initdb","msg":"initdb: error: locale \"x-icu\" has unknown language \"x-icu\"\n","pipe":"stderr","logging_pod":"test-1-initdb"}
{"level":"error","ts":"2024-11-27T20:04:41.968111086Z","msg":"Error while bootstrapping data directory","logging_pod":"test-1-initdb","error":"error while creating the PostgreSQL instance: exit status 1","stacktrace":"github.com/cloudnative-pg/machinery/pkg/log.(*logger).Error\n\tpkg/mod/github.com/cloudnative-pg/machinery@v0.0.0-20241014090714-c27747f9974b/pkg/log/log.go:125\ngithub.com/cloudnative-pg/cloudnative-pg/internal/cmd/manager/instance/initdb.initSubCommand\n\tinternal/cmd/manager/instance/initdb/cmd.go:160\ngithub.com/cloudnative-pg/cloudnative-pg/internal/cmd/manager/instance/initdb.NewCmd.func2\n\tinternal/cmd/manager/instance/initdb/cmd.go:109\ngithub.com/spf13/cobra.(*Command).execute\n\tpkg/mod/github.com/spf13/cobra@v1.8.1/command.go:985\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\tpkg/mod/github.com/spf13/cobra@v1.8.1/command.go:1117\ngithub.com/spf13/cobra.(*Command).Execute\n\tpkg/mod/github.com/spf13/cobra@v1.8.1/command.go:1041\nmain.main\n\tcmd/manager/main.go:68\nruntime.main\n\t/opt/hostedtoolcache/go/1.23.2/x64/src/runtime/proc.go:272"}
Error: error while creating the PostgreSQL instance: exit status 1
```

We're using basically the same options with a direct `dbinit` (out of Kubernetes), with no problems.

Any ideas?

_Mark

**Reply by markmaker** [reply]
Hmmm... don't understand how this is supposed to work in postgresql. It does allow `"und"` (undefined) as a _special language_ here:

https://github.com/postgres/postgres/blob/97525bc5c8ffb31475d23955d08e9ec9c1408f33/src/bin/initdb/initdb.c#L2373-L2375

but the extraction using `uloc_getLanguage()` a few lines before that seems incompatible. It ultimately goes to `_getLanguage()`, and if I understand correctly, it cuts away the "und", so it won't ever match as a _special language_:

https://github.com/unicode-org/icu/blob/e025466e3ac94bcf4a8062b98b68f4c6b7e00990/icu4c/source/common/uloc.cpp#L1208

Strange...

**Reply by markmaker** [reply]
Just reproduced out of Kubernetes. Same options produce this output:

`Using language tag "und-x-icu" for ICU locale "und-x-icu".`

However that's on PostgreSQL 16.4, not 17. Investigating...

**Reply by eshikhov** [reply]
I ran with these parameters, everything works fine
PostgreSQL 16

```
bootstrap:
    initdb:
      localeCollate: en_US.UTF8
      localeCType: en_US.UTF8
      options:
        - --locale-provider=icu 
        - --icu-locale=und
```

**Reply by markmaker** [reply]
Oh, thank you that worked! :sunglasses:

**Comment by elafarge** [comment]
I started working on the pull request, I still need to get the E2E tests to work though (didn't manage to run them locally so far) and I'll probably need to exclude postgres versions below 14 for the newly added end to end tests since ICU locales were only introduced in PG15.

https://github.com/cloudnative-pg/cloudnative-pg/pull/5368

**Comment by gbartolini** [comment]
Exactly, please ensure that this is handled correctly in case of Postgres 14 or older.

**Reply by gbartolini** [reply]
@elafarge I have created this #5386

**Reply by elafarge** [reply]
Thank you very much for your reply and for creating the issue. I'll take care of ensuring these parameters can't be set if the PG version set in the Cluster manifest is < 15, most likely this afternoon or early next week. 

Hopefully I'll also manage to get the E2E tests to run locally on kind too, and mark this PR as ready, eventually :blush:

**Comment by gbartolini** [comment, marked as answer]
This feature has been introduced in 1.25: https://cloudnative-pg.io/documentation/current/bootstrap/#passing-options-to-initdb

Closing this ticket.
